src/Locks.test.ts

```diff
@@ -0,0 +1,290 @@
+      createPresenceMessage('enter', { connectionId: '1' }),
+      createPresenceMessage('enter', { connectionId: '0' }),
+      createPresenceMessage('enter', { connectionId: '2' }),
+    ]);
+
+    context.client = client;
+    context.space = new Space('test', client);
+    context.presence = presence;
+  });
+
+  describe('acquire', () => {
+    it<SpaceTestContext>('errors if acquiring before entering the space', ({ space }) => {
+      expect(space.locks.acquire('test')).rejects.toThrowError();
+    });
+
+    it<SpaceTestContext>('enters presence with a PENDING lock request for the current member', async ({
+      space,
+      presence,
+    }) => {
+      await space.enter();
+
+      const presenceUpdate = vi.spyOn(presence, 'update');
+
+      const lockID = 'test';
+      const req = await space.locks.acquire(lockID);
+      expect(req.status).toBe('pending');
+
+      const presenceMessage = expect.objectContaining({
+        extras: {
+          locks: [
+            expect.objectContaining({
+              id: lockID,
+              status: 'pending',
+            }),
+          ],
+        },
+      });
+      expect(presenceUpdate).toHaveBeenCalledWith(presenceMessage);
+    });
+
+    it<SpaceTestContext>('includes attributes in the lock request when provided', async ({ space, presence }) => {
+      await space.enter();
+
+      const presenceUpdate = vi.spyOn(presence, 'update');
+
+      const lockID = 'test';
+      const attributes = new LockAttributes();
+      attributes.set('key1', 'foo');
+      attributes.set('key2', 'bar');
+      const req = await space.locks.acquire(lockID, { attributes });
+      expect(req.attributes).toBe(attributes);
+
+      const presenceMessage = expect.objectContaining({
+        extras: {
+          locks: [expect.objectContaining({ attributes })],
+        },
+      });
+      expect(presenceUpdate).toHaveBeenCalledWith(presenceMessage);
+    });
+
+    it<SpaceTestContext>('errors if a PENDING request already exists', async ({ space }) => {
+      await space.enter();
+
+      const lockID = 'test';
+      await space.locks.acquire(lockID);
+      expect(space.locks.acquire(lockID)).rejects.toThrowError();
+    });
+  });
+
+  describe('processPresenceMessage', () => {
+    const lockID = 'test';
+
+    const lockEvent = (member: SpaceMember, status: LockStatus) =>
+      expect.objectContaining({
+        member: member,
+        request: expect.objectContaining({ id: lockID, status }),
+      });
+
+    it<SpaceTestContext>('sets a PENDING request to LOCKED', async ({ space }) => {
+      await space.enter();
+      const member = space.members.getSelf()!;
+
+      const emitSpy = vi.spyOn(space.locks, 'emit');
+
+      const msg = Realtime.PresenceMessage.fromValues({
+        connectionId: member.connectionId,
+        extras: {
+          locks: [
+            {
+              id: lockID,
+              status: 'pending',
+              timestamp: Date.now(),
```

Comment:
What timestamp format are we planning on supporting when this goes server-side?  Would it be a JS timestamp value and we'll assume all other languages use that?

Reply:
Yeh, millisecond unix timestamp, I'll document this in the feature spec next sprint (https://ably.atlassian.net/browse/MMB-235).